### PR TITLE
JP-4041: stfitsdiff unit tests and code clean up

### PR DESCRIPTION
**Pull request by stscirij**
Added new unit tests
Cleaned up and simplified some of the code structure 
Added some more output
Fixed some bugs
Now statistics are done on numeric table elements that fail the (atol, rtol) test Report table columns that are identical
Report all extensions that are in one file but not the other

Replaces earlier PR #10006

Resolves [JP-4041](https://jira.stsci.edu/browse/JP-4041)

This PR adds some unit tests for st_fitsdiff, simplifies the code structure, changes some of the reporting and fixes a bug

## Tasks
- [ ] If you have a specific reviewer in mind, tag them.
- [x] add a build milestone, i.e. `Build 12.0` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [x] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [x] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see [changelog readme](https://github.com/spacetelescope/jwst/blob/main/changes/README.rst) for instructions) 
  - [x] update or add relevant tests
  - [ ] update relevant docstrings and / or `docs/` page
  - [ ] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

File: jwst/regtest/st_fitsdiff.py [hunk 1]
```diff
@@ -763,12 +777,13 @@ def report_data_diff():
...
 
             # Show percentage differences
-            self._writeln("\nPercentages of difference above (tolerance + threshold) ")
+            self._writeln("\nPercentages of difference above threshold")
             for tline in self.percentages.pformat():
                 self._writeln(tline)
 
         if self.diff_data is not None and not self.diff_data.identical:
             self._fileobj.write("\n")
+            self._writeln(" Data contains differences:")
```
Comment (melanieclarke) [inline jwst/regtest/st_fitsdiff.py]: There's an extra space here:

```suggestion
            self._writeln("Data contains differences:")
```

File: jwst/regtest/st_fitsdiff.py [hunk 2]
```diff
@@ -1296,32 +1309,37 @@ def get_stats_if_nonans(nonans):
                 return nonansmax, nonansmin, nonansmean
 
             # Calculate the absolute and relative differences
-            get_stats = False
-            reported = False
-            if np.issubdtype(arra.dtype, np.floating) and np.issubdtype(arrb.dtype, np.floating):
-                # Catch the case of a column full of nans
+            if np.issubdtype(arra.dtype, np.number) and np.issubdtype(arrb.dtype, np.number):
+                float_diffs = where_not_allclose(arra, arrb, rtol=self.rtol, atol=self.atol)
```
Comment (penaguerrero) [inline jwst/regtest/st_fitsdiff.py]: It seems that the indices are not needed, can you please see if you can replace the where statement (hat can take a long time and resources)? A potential way around it is to first find the nans and compare the arrays and then remove the nans and use a logical comparison like diffs = (arra != arrb)

File: jwst/regtest/st_fitsdiff.py [hunk 3]
```diff
@@ -1296,32 +1309,37 @@ def get_stats_if_nonans(nonans):
...
-                            meanr = np.mean(rel_diffs)
-                            stdr = np.std(rel_diffs)
-
+                nonansa = arra[np.isfinite(arra)]
+                nonansb = arrb[np.isfinite(arrb)]
+                arramax, arramin, arramean = get_stats_if_nonans(nonansa)
+                arrbmax, arrbmin, arrbmean = get_stats_if_nonans(nonansb)
+                finite_idx = np.isfinite(arra) & np.isfinite(arrb)
+                maxr, meanr, stdr = np.nan, np.nan, np.nan
+                r_idx = where_not_allclose(arra, arrb, self.atol, self.rtol)
```
Comment (penaguerrero) [inline jwst/regtest/st_fitsdiff.py]: this is another instance with a where statement, see the comment above

File: jwst/regtest/st_fitsdiff.py [hunk 4]
```diff
@@ -1338,131 +1356,101 @@ def get_stats_if_nonans(nonans):
...
+                for x in arrb[1:]:
+                    concatenated_b = np.concatenate((concatenated_b, x.flatten()))
+                nonans_a = concatenated_a[np.isfinite(concatenated_a)]
+                nonans_b = concatenated_b[np.isfinite(concatenated_b)]
+                arramax, arramin, arramean = get_stats_if_nonans(nonans_a)
+                arrbmax, arrbmin, arrbmean = get_stats_if_nonans(nonans_b)
+                n_fail_rtol, maxr, meanr, stdr = 0.0, 0.0, 0.0, 0.0
+                finite_idx = np.isfinite(concatenated_a) & np.isfinite(concatenated_b)
+                if finite_idx.any():
+                    r_idx = where_not_allclose(
```
Comment (penaguerrero) [inline jwst/regtest/st_fitsdiff.py:1418]: see comment above to try and replace the where statement.

**Review by penaguerrero — CHANGES_REQUESTED** [review]
There are three instances with a where statement that makes the algorithm take sometimes a bit longer than satrapy fitsdiff.

File: jwst/regtest/st_fitsdiff.py (hunk 4 above)
Comment (stscirij) [reply jwst/regtest/st_fitsdiff.py:1418]: In this case the indices r_idx are used in line 1411 below, so I'd rather keep this usage

**Review by penaguerrero — APPROVED** [review]
Thank you for making the changes, the running times have lowered. 
LGTM

File: jwst/regtest/test_stfitsdiff.py [hunk 5]
```diff
@@ -1109,44 +1131,495 @@ def test_table_pq_coltype(mock_table, fitsdiff_default_kwargs):
...
+    table_b.data["wavelength"] = np.arange(100.0) + 1.0
+    a.append(table_a)
+    b.append(table_b)
+    diff = STFITSDiff(a, b, **fitsdiff_default_kwargs)
+    report = diff.report()
+    assert "Found 100 different table data element(s)" in report
+
+
+def test_hdus_tables_misc(fitsdiff_default_kwargs):
+    a = fits.HDUList()
```
Comment (stscirij) [inline jwst/regtest/test_stfitsdiff.py:1447]: @melanieclarke I think this test covers the single NaN case you had in your regression test - the ERROR column has 5 NaNs in one  input and the report correctly displays the statistics for that column and shows the differences as NaN

Comment (melanieclarke) [reply jwst/regtest/test_stfitsdiff.py:1447]: Thanks!  This is a good idea, but I would also like to add the test case to the regression tests in `test_stfitsdiff_regtests.py`.  This is specifically to compare to the fitsdiff output for an edge case that has been a persistent problem.

File: jwst/regtest/st_fitsdiff.py [hunk 6]
```diff
@@ -1127,10 +1143,6 @@ def __init__(
             names=(
                 "col_name",
                 "dtype",
-                "abs_diffs",
-                "abs_max",
-                "abs_mean",
-                "abs_std",
                 "rel_diffs",
```
Comment (melanieclarke) [inline jwst/regtest/st_fitsdiff.py:1146]: Question - what is this quantity reporting?  For an example table I have, it reports 995 pixels failed the (atol, rtol) test, all in the FLUX column, then it reports rel_diffs for FLUX as 1918 in the table.  I would have expected that value to be the number of different pixels, so 995?

Comment (stscirij) [reply jwst/regtest/st_fitsdiff.py:1146]: > Question - what is this quantity reporting? For an example table I have, it reports 995 pixels failed the (atol, rtol) test, all in the FLUX column, then it reports rel_diffs for FLUX as 1918 in the table. I would have expected that value to be the number of different pixels, so 995?

This is fixed in [8771c9d](https://github.com/spacetelescope/jwst/pull/10065/commits/8771c9d44892ea748c434ef65c6c054d6ae18f49)

File: jwst/regtest/st_fitsdiff.py (hunk 1 above)
Comment (stscirij) [reply jwst/regtest/st_fitsdiff.py]: Fixed in [8771c9d](https://github.com/spacetelescope/jwst/pull/10065/commits/8771c9d44892ea748c434ef65c6c054d6ae18f49)

File: jwst/regtest/st_fitsdiff.py (hunk 6 above)
Comment (stscirij) [reply jwst/regtest/st_fitsdiff.py:1146]: I also added a new regtest and uploaded the files to artifactory

Comment (melanieclarke) [reply jwst/regtest/st_fitsdiff.py:1146]: Looks good now, thanks!

**Review by melanieclarke — APPROVED** [review]
I think all comments are addressed now, and the reports and tests look good. Thanks very much for all your hard work on this one!

**Comment by melanieclarke** [comment]
There are some unrelated failures in the CI tests.  Since this PR isn't urgent, let's wait and see if they get resolved on main, then we can get this merged.

**Comment by melanieclarke** [comment]
CI is passing now; I will merge.
